src/loop_in_chunks.cpp

```diff
@@ -180,9 +180,51 @@ integral = f_i [-w0^3/3 + (i+1) w0^2/2
 ****************************************************************************/
 
 using namespace std;
+typedef complex<double> cdouble;
 
 namespace meep {
 
+/***************************************************************/
+/* utility routine designed to be called by chunkloop functions*/
+/* for fetching values of field components at grid points,     */
+/* taking into account the complications of symmetry and       */
+/* yee-grid averaging.                                         */
+/* The first 5 arguments are the arguments of the same name    */
+/* passed to the chunkloop function.                           */
+/* The 6th argument is the integer idx defined inside the body */
+/* of a LOOP_OVER_IVECS.                                       */
+/* components is a caller-created list of components; the      */
+/* return value is an array of the same length storing the     */
+/* values of those components at the grid point.               */
+/***************************************************************/
+vector<complex<double> > get_field_components(fields_chunk *fc, component cgrid,
+                                              complex<double> shift_phase,
+                                              const symmetry &S, int sn,
+                                              int idx, vector<component> components)
```

Comment:
This really should be refactored into two routines, one since the call to `S.transform`, `S.phase_shift`, and `yee2cent_offsets` can happen once outside the ivec loop.   That is, there should be one routine that sets up a data structure or class for field computations (outside the loop) and one that applies it to `fc->f` (inside the loop).

Reply:
Yes, I considered this optimization, but it would make the code harder to read and would yield only modest speedup, at least for field components. (This optimization would be more essential for permittivity and permeability, which is why I omitted them here.) 

On the spectrum from "easy to read but not speed-optimized" versus "replete with specialized tweaks and accelerations that obfuscate what's going on," meep is already massively weighted toward the latter extreme, and the whole point of this documentation and this utility routine are to lower the entry barrier for new developers to become familiar with basic functionality. If there comes a time when the slightly suboptimal implementation of `get_field_components` actually slows something down in real life, I'll happily implement a more complicated speed-optimized version.